ethcore/src/blockchain/blockchain.rs

```diff
@@ -843,6 +832,80 @@ impl BlockChain {
 		}
 	}
 
+	/// Update the best ancient block to the given hash, after checking that
+	/// it's directly linked to the currently known best ancient block
+	pub fn update_best_ancient_block(&self, hash: &H256) {
+		// Get the block view of the next ancient block (it must
+		// be in DB at this point)
+		let block_view = match self.block(hash) {
+			Some(v) => v,
+			None => return,
+		};
+
+		// So that `best_ancient_block` gets unlocked before calling
+		// `set_best_ancient_block`
+		{
+			// Get the target hash ; if there are no ancient block,
+			// it means that the chain is already fully linked
+			let best_ancient_block = self.best_ancient_block.read();
+			let cur_ancient_block = match *best_ancient_block {
+				Some(ref b) => b,
+				None => return,
+			};
+			let target_hash = cur_ancient_block.hash;
+
+			// Ensure that the new best ancient block is after the current one
+			if block_view.number() <= cur_ancient_block.number {
+				return;
+			}
+
+			let mut block_hash = *hash;
+			let mut is_linked = false;
+
+			loop {
```

Comment:
Actually, it prevents the `update_best_ancient_block` function to be called twice in parallel, which could break the wanted behavior.

Reply:
It doesn't really. `RwLock::read` can be acquired from multiple threads, although it might deadlock if called from the same thread recursively.
Calling other functions dependent on `sefl` while holding a lock is imho asking for troubles, especially if the code is not thoroughly tested. Another person will try to access the same lock in some other function and it will break.

Reply:
Hm, right. So should I use a `write` lock? The only function called here is `block_details` which we are sure doesn't use any lock.

Reply:
I believe there is a logic error here (or in `insert_unordered_block`).

`insert_unordered_block` updates `best_ancient_block`, but does not update `block_details`

https://github.com/paritytech/parity-ethereum/blob/1b1941a896c8485a05fa3e9ffe8b251b498eb541/ethcore/src/blockchain/blockchain.rs#L782-L795

it updates only `pending_block_details`

https://github.com/paritytech/parity-ethereum/blob/1b1941a896c8485a05fa3e9ffe8b251b498eb541/ethcore/src/blockchain/blockchain.rs#L1092-L1098

`block_details` are updated only in `commit` function

https://github.com/paritytech/parity-ethereum/blob/1b1941a896c8485a05fa3e9ffe8b251b498eb541/ethcore/src/blockchain/blockchain.rs#L1127

So there will be a race condition if this function is called between `insert` and `commit`

Reply:
^^ @ngotchac is this addressed?